PyInstaller/utils/win32/winutils.py

```diff
@@ -47,6 +51,21 @@ def get_pe_file_machine_type(filename):
         return pe.FILE_HEADER.Machine
 
 
+def rewrite_exe(exe_path, data, retries=20):
+    """
+    This function writes `data` to the `exe_path` while catching OSError exceptions until `retries` are reached.
+    """
+    for retry in range(retries):
+        try:
+            with open(exe_path, 'wb') as fp:
+                fp.write(data)
+            return
+        except OSError:
```

Comment:
Can we not be more explicit here?
```suggestion
        except PermissionError:
```
`OSError` has a lot of subclasses which aren't applicable here:
```python
>>> OSError.__subclasses__()
[<class 'ConnectionError'>, <class 'BlockingIOError'>, <class 'ChildProcessError'>, <class 'FileExistsError'>, <class 'FileNotFoundError'>, <class 'IsADirectoryError'>, <class 'NotADirectoryError'>, <class 'InterruptedError'>, <class 'PermissionError'>, <class 'ProcessLookupError'>, <class 'TimeoutError'>, <class 'io.UnsupportedOperation'>, <class 'signal.itimer_error'>, <class 'socket.herror'>, <class 'socket.gaierror'>, <class 'ssl.SSLError'>, <class 'shutil.Error'>, <class 'shutil.SpecialFileError'>, <class 'shutil.ExecError'>, <class 'shutil.ReadError'>]
```

Reply:
Assuming you can use `PermissionError()`, mind also changing that `raise OSError(...)` below to a `raise PermissionError(...)` too.

Reply:
The exact error that I observed what feels like 100 times is: OSError: [Errno 22] Invalid argument: '<path>.exe.notanexecutable' which doesn't seem to match any of the subclasses as far as I'm aware. 

I'm not lightly catching all OSError exceptions here either that's why I only retry a few times before continuing to raise it.

Reply:
> The exact error that I observed what feels like 100 times is: OSError: [Errno 22] Invalid argument: '.exe.notanexecutable' which doesn't seem to match any of the subclasses as far as I'm aware.

Out of curiosity, is this on a local filesystem (with a local AV scanner), or is it some sort of shared/remote drive (with non-local scanner)?

If we wanted to limit the exceptions, we could probably do for `PermissionError` and `OSError` with `errno` 22 (˙`EINVAL`).

That said, the retry function should re-raise the last original exception, not throw its own.

Reply:
> > The exact error that I observed what feels like 100 times is: OSError: [Errno 22] Invalid argument: '.exe.notanexecutable' which doesn't seem to match any of the subclasses as far as I'm aware.
> 
> Out of curiosity, is this on a local filesystem (with a local AV scanner), or is it some sort of shared/remote drive (with non-local scanner)?
In one case it's a thing called Crowdstrike mandated by my employer on my work windows laptop. 
In the other case it's a Windows VM with the same tool installed where a (Windows) Docker container is run working in a mounted workspace from the Host VM. 
But turning off Crowdstrike for a folder on my machine (the dist folder) didn't change anything so I'm not sure it's actually caused by this. 
> 
> If we wanted to limit the exceptions, we could probably do for `PermissionError` and `OSError` with `errno` 22 (˙`EINVAL`).
> 
Sounds good to me
> That said, the retry function should re-raise the last original exception, not throw its own.
Usually I'd agree but in this case the error message (at least for the OSError) is really confusing. How is a user of pyinstaller supposed to understand what the "invalid parameter" is when the parameter is literally a valid path on the FS created by pyinstaller.